toolchain/common/enum_base.h

```diff
@@ -0,0 +1,179 @@
+// support `case MyEnum::Name:`.
+//
+// It is specifically designed to compose with X-MACRO style `.def` files that
+// stamp out all the enumerators.
+//
+// It also supports some opt-in APIs that classes can enable by `using` the
+// names to make them public: `AsInt` and `FromInt` allow converting to and from
+// the underlying type of the enumerator.
+//
+// Users must be in the `Carbon` namespace and should look like the following.
+// In `my_kind.h`:
+//   ```
+//   CARBON_DEFINE_RAW_ENUM_CLASS(MyKind, uint8_t){
+//   #define CARBON_MY_KIND(Name) CARBON_RAW_ENUM_ENUMERATOR(Name)
+//   #include "toolchain/.../my_kind.def"
+//   };
+//
+//   class MyKind : public CARBON_ENUM_BASE(MyKind) {
+//    public:
+//   #define CARBON_MY_KIND(Name) CARBON_ENUM_CONSTANT_DECLARATION(Name)
+//   #include "toolchain/.../my_kind.def"
+//   };
+//
+//   #define CARBON_MY_KIND(Name) CARBON_ENUM_CONSTANT_DEFINITION(MyKind, Name)
+//   #include "toolchain/.../my_kind.def"
+//   ```
+// In `my_kind.cpp`:
+//   ```
+//   CARBON_DEFINE_ENUM_CLASS_NAMES(MyKind) = {
+//   #define CARBON_MY_KIND(Name) CARBON_ENUM_CLASS_NAME_STRING(Name)
+//   #include "toolchain/.../my_kind.def"
+//   };
+//   ```
+template <typename DerivedT, typename EnumT>
+class EnumBase {
+ protected:
+  // An alias for the raw enum type. This is an implementation detail and
+  // shouldn't be used, but we need it for a signature so it is declared early.
+  using RawEnumType = EnumT;
+
+ public:
+  using EnumType = DerivedT;
+  using UnderlyingType = std::underlying_type_t<RawEnumType>;
+
+  // Enable conversion to the raw enum type, including in a `constexpr` context,
+  // to enable comparisons and usage in `switch` and `case`. The enum type
+  // remains an implementation detail and nothing else should be using this
+  // function.
+  //
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  constexpr operator RawEnumType() const { return value_; }
+
+  // Conversion to bool is deleted to prevent direct use in an `if` condition
+  // instead of comparing with another value.
+  explicit operator bool() const = delete;
+
+  // Returns the name of this value.
+  //
+  // This method will be automatically defined using the static `names` string
+  // table in the base class, which is in turn will be populated for each
+  // derived type using the macro helpers in this file.
+  [[nodiscard]] auto name() const -> llvm::StringRef;
+
+  // Prints this value using its name.
+  void Print(llvm::raw_ostream& out) const {
+    out << reinterpret_cast<const EnumType*>(this)->name();
+  }
+
+ protected:
+  // The default constructor is explicitly defaulted (and constexpr) as a
+  // protected constructor to allow derived classes to be constructed but not
+  // the base itself. This should only be used in the `Create` function below.
+  constexpr EnumBase() = default;
+
+  // Create an instance from the raw enumerator. This should only be used to
+  // initialize the named constants in the derived type.
```

Comment:
```suggestion
  // Create an instance from the raw enumerator, for internal use.
```

"This should only..."

But it's also used for FromInt.